debian/cudf-check.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.38.2.
.TH CUDF-CHECK "1" "June 2010" "cudf-check 0.5" "User Commands"
.SH NAME
cudf-check \- manipulate CUDF documents
.SH SYNOPSIS
.B cudf-check
[\fIOPTION\fR...]
.SH DESCRIPTION
.SS "In particular:"
.TP
cudf\-check \fB\-cudf\fR FILE
validate CUDF
.TP
cudf\-check \fB\-cudf\fR FILE \fB\-sol\fR FILE
validate CUDF and its solution
.TP
cudf\-check \fB\-univ\fR FILE
validate package universe (no request)
.SH OPTIONS
.HP
\fB\-cudf\fR parse the given CUDF (universe + request)
.HP
\fB\-univ\fR parse the given package universe
.HP
\fB\-sol\fR parse the given problem solution
.HP
\fB\-dump\fR dump parse results to standard output
.TP
\fB\-help\fR
Display this list of options
.TP
\fB\-\-help\fR
Display this list of options


ChangeLog
2010-06-25  Stefano Zacchiroli  <zack@pps.jussieu.fr>

	* ALL bug fix: do not enforce Keep on uninstalled packages
	  Thanks Ralf Treinen for the patch! (Closes: #266)
	* ALL bug fix: nicer error message for unknown packages found in
	  solution (Closes: #267)
	* ALL bug fix: avoid pretty printing some malformed vpkgformulae
	  (Closes: #272)
	* ALL bug fix: check for duplicate property in 822 stanzas
	  (Closes: #269)
	* ALL bug fix: comments no longer split stanzas in two
	  (Closes: #250)
	* ALL bug fix: perform string escape when pretty printing type
	  declarations (Closes: #271)
	* ALL add several test cases to the test suite
	* ALL Development release 0.5.98

2010-06-03  Stefano Zacchiroli  <zack@pps.jussieu.fr>

	* ALL bug fix: optional/mandatory extra property management
	* c-lib bug fix: add handling of vpkgformula extra properties
	* ALL bug fix: allow empty vpkglist as default values
	* ALL bug fix: do not accept trivial formulae as v(eq)pkglist
	* ALL Development release 0.5.97

2010-06-01  Stefano Zacchiroli  <zack@pps.jussieu.fr>

	* c-lib add structured accessors to request fields
	  (patch by Gustavo Gutierrez, thanks!)
	* doc better explain the role of load_solution
	* ALL simpler API for solutions: cudf_load_solution_from_file
	* ALL bug fix: avoid losing ?typedecl in convenience functions
	* c-lib add bindings for cudf_load_solution_from_file
	* backward-compatibility enabling compilation with OCaml 3.10.2
	* ALL bug fix: support heading '+' when parsing posint
	* ALL Development release 0.5.96

2009-12-18  Stefano Zacchiroli  <zack@pps.jussieu.fr>

	* c-lib Remove the need of OCaml devel stuff to use the C API
	* c-lib Provide bindings to OCaml native code, if available
	* c-lib Hide OCaml "value", it is way more safe (API change)
	* c-lib O(n^2) -> O(n) speed-up in package list conversion
	* ALL Development release 0.5.94

2009-11-25  Stefano Zacchiroli  <zack@pps.jussieu.fr>

	* ALL Porting to CUDF 2.0
	* ALL Development release 0.5.92

2009-02-24  Stefano Zacchiroli  <zack@pps.jussieu.fr>

	* ALL First public release, version 0.4



cudf-check.ml
(*****************************************************************************)
(*  libCUDF - CUDF (Common Upgrade Description Format) manipulation library  *)
(*  Copyright (C) 2009  Stefano Zacchiroli <zack@pps.jussieu.fr>             *)
(*                                                                           *)
(*  This library is free software: you can redistribute it and/or modify     *)
(*  it under the terms of the GNU Lesser General Public License as           *)
(*  published by the Free Software Foundation, either version 3 of the       *)
(*  License, or (at your option) any later version.  A special linking       *)
(*  exception to the GNU Lesser General Public License applies to this       *)
(*  library, see the COPYING file for more information.                      *)
(*****************************************************************************)

open ExtLib
open Printf

open Cudf
open Cudf_checker

let cudf_arg = ref ""
let univ_arg = ref ""
let sol_arg = ref ""
let dump_arg = ref false

let cudf = ref None
let univ = ref None
let sol = ref None

let arg_spec = [
  "-cudf", Arg.Set_string cudf_arg,
    "parse the given CUDF (universe + request)" ;
  "-univ", Arg.Set_string univ_arg, "parse the given package universe" ;
  "-sol", Arg.Set_string sol_arg, "parse the given problem solution" ;
  "-dump", Arg.Set dump_arg, "dump parse results to standard output" ;
]

let usage_msg =
"Usage: cudf-check [OPTION...]
In particular:
  cudf-check -cudf FILE               validate CUDF
  cudf-check -cudf FILE -sol FILE     validate CUDF and its solution
  cudf-check -univ FILE               validate package universe (no request)
Options:"

let die_usage () = Arg.usage arg_spec usage_msg ; exit (-2)

let print_inst_info inst =
  match is_consistent inst with
    | true, _ -> printf "installation status consistent\n%!"
    | false, Some r ->
	printf "installation status inconsistent (reason: %s)\n%!"
	  (explain_reason (r :> bad_solution_reason))
    | _ -> assert false

let print_cudf (pre, univ, req) =
  if !dump_arg then begin
    let pre' = Option.default Cudf.default_preamble pre in
    print_endline (Cudf_printer.string_of_cudf (pre', univ, req))
  end

let print_univ univ =
  if !dump_arg then
    print_endline (Cudf_printer.string_of_universe univ)

let print_sol_info inst sol =
  match is_solution inst sol with
    | true, _ -> printf "is_solution: true\n%!"
    | false, rs ->
	printf "is_solution: false (reason: %s)\n%!"
	  (String.concat "; " (List.map explain_reason rs))

let pp_loc (start_pos, end_pos) =
  let pp_lpos { Lexing.pos_fname = _fname;
		pos_lnum = lnum; pos_bol = bol; pos_cnum = cnum } =
    sprintf "%d" lnum in
  sprintf "line: %s" (pp_lpos start_pos)

let main () =
  let load_univ p = 
    let pre,univ,req = Cudf_parser.load p in
    univ in
  let fail_parse source msg loc =
    eprintf "Error while parsing %s: %s\n" source msg ;
    eprintf "Location: %s\n%!" (pp_loc loc) ;
    exit 1
  in
  if !cudf_arg <> "" then begin
    try
      let p = Cudf_parser.from_in_channel (open_in !cudf_arg) in
      eprintf "loading CUDF ...\n%!";
      (match Cudf_parser.load p with
	 | pre, univ, None ->
             eprintf "Error: missing request description item.\n%!";
	     exit (-1)
	 | pre, univ, Some req -> cudf := Some (pre, univ, req))
    with
      | Cudf_parser.Parse_error (msg, loc) -> fail_parse "CUDF" msg loc
      | Cudf.Constraint_violation _ as exn ->
          eprintf "Error while loading CUDF from %s: %s\n%!"
            !cudf_arg (Printexc.to_string exn);
          exit (-1)
  end;
  if !univ_arg <> "" then begin
    try
      let p = Cudf_parser.from_in_channel (open_in !univ_arg) in
      eprintf "loading package universe ...\n%!";
      univ := Some (load_univ p)
    with
      | Cudf_parser.Parse_error (msg, loc) -> fail_parse "universe" msg loc
      | Cudf.Constraint_violation _ as exn ->
          eprintf "Error while loading universe from %s: %s\n%!" 
            !univ_arg (Printexc.to_string exn);
          exit (-1)
  end;
  if !sol_arg <> "" then begin
    try
      let p = Cudf_parser.from_in_channel (open_in !sol_arg) in
      eprintf "loading solution ...\n%!";
      sol := Some (load_univ p)
    with
      | Cudf_parser.Parse_error (msg, loc) -> fail_parse "solution" msg loc
      | Cudf.Constraint_violation _ as exn ->
          eprintf "Error while loading solution from %s: %s\n%!"
            !sol_arg (Printexc.to_string exn);
          exit (-1)
  end;
  match !cudf, !univ, !sol with
  | Some (pre,univ,req), None, None ->
      print_inst_info univ;
      print_cudf (pre,univ,req)
  | Some (pre,univ,req), None, Some sol ->
      print_inst_info univ;
      print_sol_info (univ,req) sol;
      print_cudf (pre,univ,req)
  | None, Some univ, None ->
      print_inst_info univ;
      print_univ univ
  | _ -> die_usage ()

let _ = 
  Arg.parse arg_spec ignore usage_msg;
  main()
